File: card-core/src/main/java/com/jean202/cardmizer/core/application/UpdatePriorityService.java
package com.jean202.cardmizer.core.application;

import com.jean202.cardmizer.core.domain.Card;
import com.jean202.cardmizer.core.domain.CardId;
import com.jean202.cardmizer.core.domain.PriorityStrategy;
import com.jean202.cardmizer.core.port.in.UpdatePriorityUseCase;
import com.jean202.cardmizer.core.port.out.LoadCardCatalogPort;
import com.jean202.cardmizer.core.port.out.UpdateCardPriorityPort;
import java.util.LinkedHashSet;
import java.util.List;
import java.util.Objects;
import java.util.Set;
import java.util.stream.Collectors;

public class UpdatePriorityService implements UpdatePriorityUseCase {
    private final LoadCardCatalogPort loadCardCatalogPort;
    private final UpdateCardPriorityPort updateCardPriorityPort;

    public UpdatePriorityService(
            LoadCardCatalogPort loadCardCatalogPort,
            UpdateCardPriorityPort updateCardPriorityPort
    ) {
        this.loadCardCatalogPort = Objects.requireNonNull(loadCardCatalogPort, "loadCardCatalogPort must not be null");
        this.updateCardPriorityPort = Objects.requireNonNull(
                updateCardPriorityPort,
                "updateCardPriorityPort must not be null"
        );
    }

    @Override
    public void update(PriorityStrategy priorityStrategy) {
        Objects.requireNonNull(priorityStrategy, "priorityStrategy must not be null");

        List<CardId> orderedCardIds = priorityStrategy.orderedCardIds();
        Set<CardId> uniqueOrderedCardIds = new LinkedHashSet<>(orderedCardIds);
        if (uniqueOrderedCardIds.size() != orderedCardIds.size()) {
            throw new IllegalArgumentException("Priority strategy must not contain duplicate card ids");
        }

        Set<CardId> configuredCardIds = loadCardCatalogPort.loadAll().stream()
                .map(Card::id)
                .collect(Collectors.toCollection(LinkedHashSet::new));

        if (!configuredCardIds.equals(uniqueOrderedCardIds)) {
            Set<CardId> missingCardIds = new LinkedHashSet<>(configuredCardIds);
            missingCardIds.removeAll(uniqueOrderedCardIds);

            Set<CardId> unknownCardIds = new LinkedHashSet<>(uniqueOrderedCardIds);
            unknownCardIds.removeAll(configuredCardIds);

            String detail = buildMismatchDetail(missingCardIds, unknownCardIds);
            throw new IllegalArgumentException("Priority strategy must include every configured card exactly once" + detail);
        }

        updateCardPriorityPort.update(priorityStrategy);
    }

    private String buildMismatchDetail(Set<CardId> missingCardIds, Set<CardId> unknownCardIds) {
        StringBuilder builder = new StringBuilder();
        if (!missingCardIds.isEmpty()) {
            builder.append(" (missing: ")
                    .append(formatCardIds(missingCardIds))
                    .append(')');
        }
        if (!unknownCardIds.isEmpty()) {
            builder.append(builder.isEmpty() ? " (" : " ")
                    .append("unknown: ")
                    .append(formatCardIds(unknownCardIds))
                    .append(')');
        }
        return builder.toString();
    }

    private String formatCardIds(Set<CardId> cardIds) {
        return cardIds.stream()
                .map(CardId::value)
                .collect(Collectors.joining(", "));
    }
}

File: card-core/src/main/kotlin/com/jean202/cardmizer/core/application/UpdatePriorityService.kt
package com.jean202.cardmizer.core.application

import com.jean202.cardmizer.core.domain.CardId
import com.jean202.cardmizer.core.domain.PriorityStrategy
import com.jean202.cardmizer.core.port.`in`.UpdatePriorityUseCase
import com.jean202.cardmizer.core.port.out.LoadCardCatalogPort
import com.jean202.cardmizer.core.port.out.UpdateCardPriorityPort

class UpdatePriorityService(
    private val loadCardCatalogPort: LoadCardCatalogPort,
    private val updateCardPriorityPort: UpdateCardPriorityPort,
) : UpdatePriorityUseCase {

    override fun update(priorityStrategy: PriorityStrategy) {
        val orderedCardIds = priorityStrategy.orderedCardIds
        val uniqueOrderedCardIds = orderedCardIds.toLinkedHashSet()
        require(uniqueOrderedCardIds.size == orderedCardIds.size) {
            "Priority strategy must not contain duplicate card ids"
        }

        val configuredCardIds = loadCardCatalogPort.loadAll().map { it.id }.toLinkedHashSet()
        if (configuredCardIds != uniqueOrderedCardIds) {
            val missingCardIds = configuredCardIds - uniqueOrderedCardIds
            val unknownCardIds = uniqueOrderedCardIds - configuredCardIds
            val detail = buildMismatchDetail(missingCardIds, unknownCardIds)
            throw IllegalArgumentException("Priority strategy must include every configured card exactly once$detail")
        }

        updateCardPriorityPort.update(priorityStrategy)
    }

    private fun buildMismatchDetail(missingCardIds: Set<CardId>, unknownCardIds: Set<CardId>): String {
        val parts = buildList {
            if (missingCardIds.isNotEmpty()) add("missing: ${formatCardIds(missingCardIds)}")
            if (unknownCardIds.isNotEmpty()) add("unknown: ${formatCardIds(unknownCardIds)}")
        }
        return if (parts.isEmpty()) "" else " (${parts.joinToString(" ")})"
    }

    private fun formatCardIds(cardIds: Set<CardId>) = cardIds.joinToString(", ") { it.value }

    private fun <T> List<T>.toLinkedHashSet(): LinkedHashSet<T> = LinkedHashSet(this)
}
